### PR TITLE
Update test_scrollcontainer.py

### PR DESCRIPTION
**Pull request by Hassan-Mallah**
## PR Checklist:

- [ ] All new features have been tested
- [ ] All new features have been documented
- [ ] I have read the **CONTRIBUTING.md** file
- [ ] I will abide by the code of conduct

### CODE REVIEW

**Comment by Hassan-Mallah** [comment]
How do we test void functions?
Specially refresh, what values are changing?

**Comment by Hassan-Mallah** [comment]
I'll work on this one now.

I'll do one test each time :)

**Comment by Hassan-Mallah** [comment]
@freakboy3742 is this one ready, or what do i need to change?
Thnks

**Comment by Hassan-Mallah** [comment]
Thank you for the feedback. I hope i got it :)

what do you think of this test:
`

        self.assertIsNone(self.sc.window)

        new_content = toga.Box(style=TestStyle(), factory=toga_dummy.factory)
        self.sc.content = new_content
        self.assertActionPerformed(self.sc, 'set content')

        self.sc.window = toga.window
        self.assertEqual(self.sc.window, toga.window)

        self.assertValueSet(self.sc, 'window', toga.window)`

I added test for null, and set value for content with test as well.
Now it's better coverage, cuz i wasn't setting content.

Or should i do 2 tests: 'test set window' and 'test set window with content'

Cheers

**Comment by freakboy3742** [comment]
@Hassan-Mallah Close, but not quite.

You've got the right idea that you need to verify the behavior by setting `.window`, not invoking _set_window directly. And you're verifying a precondition around the window of the scroll container.

However, you're not verifying the "edge case" of _set_content - what happens if the scroll container *doesn't* have pre-existing content? In that case, you need to verify that something *doesn't* happen.

I'd probably also update the test name - it's not strictly a test of `_set_window` anymore.

You're also correct that 2 tests may be better here. In an ideal world, every test should be "something we can assert is behaving correctly". Verifying that correctness may require multiple actual assert calls, but there should be a single "fact" known after running any single test (e.g., "if the scrollcontainer has content, that content has it's window set").

**Comment by Hassan-Mallah** [comment]
Hey,

Thank you for the feedback. 
1. I made them 2 tests, with and without content.
2. I tested _children, is it extra?
3. I used assertRaises with None sc

is it too many changes :D 
cheers
`

        def test__set_window_without_content(self):
                self.assertIsNotNone(self.sc)
                self.assertIsNone(self.sc.window)

                self.sc.window = toga.window
                self.assertEqual(self.sc.window, toga.window)

                self.assertIsNone(self.sc._children)
                self.assertValueSet(self.sc, 'window', toga.window)

                self.sc = None
                with self.assertRaises(AttributeError):
                    self.sc.window = toga.window

        def test__set_window_with_Content(self):
                self.assertIsNone(self.sc.window)

                new_content = toga.Box(style=TestStyle(), factory=toga_dummy.factory)
                self.sc.content = new_content
                self.assertActionPerformed(self.sc, 'set content')

                self.sc.window = toga.window
                self.assertEqual(self.sc.window, toga.window)

                self.assertValueSet(self.sc, 'window', toga.window)
`

**Comment by freakboy3742** [comment]
This is getting closer - however, it's very hard to provide review notes when you've submitted the changes as a comment. If you update your PR, I can make comments against the specific lines of code that have issues. 

My comment about test naming is that this isn't a test of `_set_window` any more. Increasing coverage of `_set_window` is a required side effect of what you're actually doing - testing that you can alter the window attribute on a scroll container. Test names like `test_assign_to_window_with_content` and `test_assign_to_window_without_content` make more sense here, because they're not tied to the specific method where you're trying to improve coverage - they describe the general thing that *causes* that coverage to increase.

Regardless - it's always a good idea to add code comments describing what you're doing, and why. I don't mean 
```
# Set window to None
widget.window = None
```
- that would be self-evident. I mean describing the *purpose* of each line (or group of lines)
```
# Ensure the widget is not assigned to a window
widget.window = None
```

**Comment by Hassan-Mallah** [comment]
Hey,

It's been 1 week since last time :)

I submitted my code and added some comments as well, thank you for the feedback.

cheers

File: src/core/tests/widgets/test_scrollcontainer.py [hunk 1]
```diff
@@ -45,3 +45,33 @@ def test_horizontal_property(self):
...
+        # children remains None when window is set
+        self.assertIsNone(self.sc._children)
+        self.assertValueSet(self.sc, 'window', toga.window)
+
+        # test that error is raised when setting a window to empty ScrollContainer
+        self.sc = None
+        with self.assertRaises(AttributeError):
+            self.sc.window = toga.window
+
+    def test__set_window_with_Content(self):
```
Comment (freakboy3742) [inline src/core/tests/widgets/test_scrollcontainer.py:66]: Again; comments on test naming haven't been addressed.

**Comment by Hassan-Mallah** [comment]
Hey,
useful feedback, and some sill mistakes from my side :)
"comments about test naming" You wanted me to change set to assign, that's all?

cheers

**Comment by freakboy3742** [comment]
@Hassan-Mallah It's no problem having multiple tests in a single PR - in fact, if the tests are all related, it's a good idea to have them all in one PR. However, they do need to be named accurately. In this case, the test *isn't* a test of `_set_window`, so naming it `test__set_window_...` is misleading. The test name should reflect what the test is actually testing. `test_assign_...` is a better name in this context because thats what you're actually doing.

**Comment by Hassan-Mallah** [comment]
I understand, thank you for explanation.
When there is more than one test, it'll take me more time to figure it out.
With more experience, i'll be doing more tests in one PR :)

What's left to do with this test?

Cheers

**Comment by freakboy3742** [comment]
Well, I'd like to see the other tests restored. The purpose of this process isn't to make "a test" - it's to improve testing of a feature. That will almost certainly involve multiple actual tests. You *were* on the right track previously; cutting this back to 1 test was the wrong direction in this case.

**Comment by Hassan-Mallah** [comment]
Hey,
I committed the 2nd test.
cheers

**Comment by freakboy3742** [comment]
Closing this PR due to lack of activity.
